### PR TITLE
Fix DefaultAnalyzerAssemblyLoader

### LINKED ISSUE #56254: Diagnostic analyzer runner is unavailable on NetCore OOP
**Version Used**: Version 17.0.0 Preview 5.0 [31706.66.main]

**Steps to Reproduce**:

1.  Turn on NetCore OOP
2.  Open a test project

**Expected Behavior**:
No error

**Actual Behavior**:
MicrosoftTeams-image

### PR DESCRIPTION
**Pull request by genlu**
Explicitly reference the AssemblyLoadContext where compiler is being loaded into instead of relying on fallback to AssemblyLoadConext.Default.

In Servicehub .Net Core host, each service is assigned a separate ALC (Roslyn is one of them), whereas the Default ALC is used by Servicehub itself.
See [here ](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/DevCore?path=/src/clr/hosts/Microsoft.ServiceHub.HostLib/IsolationServiceManager.cs&_a=contents&version=GBdevelop)for more details on the Servicehub implementation.

Fix #56254

### CODE REVIEW

**Comment by genlu** [comment]
@jaredpar @RikkiGibson @tmat

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs:123]: It might be best to also change the comment on line 131 to refer to "default ALC" instead of "parent ALC".

**Review by RikkiGibson — COMMENTED** [review]
The change seems fine but it would be very desirable to have a unit test which fails when the change is not applied. I think we should make a new test similar to `LoadWithDependency`:

https://github.com/dotnet/roslyn/blob/2aa0ac8c481e632e0aead4259db50f119841a979/src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs#L43-L62

I think it would be adequate to just create some helper method which performs similar logic to this existing test: create an analyzer loader and verify that it can load an analyzer which depends on the compiler.

Then all we have to do is load the test class into a new ALC--some concrete subclass of ALC that is declared just for the purposes of the test--get the test class out of the loaded assembly, and then call the helper method. I think we'd have to call via reflection.

I'd be happy to help with creating this test if you are short on time. But I do want to make sure that our automation can detect when we have broken this scenario if possible.

**Review by jaredpar — CHANGES_REQUESTED** [review]
We definitely need a unit test here. This particular scenario isn't covered in any of the documentation on how ALC works. We need a test, and likely several, to verify this is going to have the expected behavior here. 

I'm fairly skeptical this is going to work when the default ALC != to the compiler ALC. The reason being that this change will result in the compiler and analyzer ALC having different copies of the assembly where our analyzer / compiler relationship requires them to share one. Essentially the compiler assemblies need to be loaded into the default ALC then shared with the analyzer ALC . This change should result in them getting separate instancese

I also think we need to sync up with the ALC runtime team and make sure that the behavior we see here is expected. Essentially is it okay for ALC to communicate in this way: essentially sharing dependencies

**Comment by jaredpar** [comment]
I've confirmed with the runtime team that this will not work as this PR expects it to. ALC are not hierarchical.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs
Comment (genlu) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs]: Fixed

**Comment by genlu** [comment]
@jaredpar @RikkiGibson I have made the change as discussed offline and added a test.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs
Comment (jaredpar) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs:135]:
```suggestion
                    _compilerLoadContext.LoadFromAssemblyName(assemblyName);
```

This also needs to fall back to the compiler ALC

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs:74]: It's not clear to me how the 'static readonly' could lead to problems, given that there's 1 static field per assembly, essentially, but I think @jaredpar's suggestion was to eliminate the static field at the same time as introducing an instance field to the `DirectoryLoadContext`.

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:445]: Neat, for some reason I had thought this class is abstract.

**Review by jaredpar — CHANGES_REQUESTED** [review]

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:453]: nit: seems more uniform to `var sb = new StringBuilder();`

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:458]:
```suggestion
        // This helper does the same thing as in `AssemblyLoading_AnalyzerReferencesSystemCollectionsImmutable_01` test above except the assertions.
```

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs:73]:
```suggestion
        // This is the default context where compiler (and some of its dependencies) are being loaded into, which might be different from AssemblyLoadContext.Default.
```

**Comment by jaredpar** [comment]
> I'm fairly skeptical this is going to work when the default ALC != to the compiler ALC. The reason being that this change will result in the compiler and analyzer ALC having different copies of the assembly where our analyzer / compiler relationship requires them to share one. Essentially the compiler assemblies need to be loaded into the default ALC then shared with the analyzer ALC . This change should result in them getting separate instancese

Wanted to follow up on this comment for anyone reading the PR (team already had a chat about this). 

My understanding of how ALC work here was incorrect. I mistakenly believed that the ALC infrastructure would always cache the `Assembly` returned from `Load` in the ALC instance. Hence my statement that this wouldn't work because it created a different `Assembly` in each ALC here. In reality though `Assembly` are tied to ALC and understand which one they belong to. Hence the change is completely safe. 

In fact you get the same behavior in `AssemblyLoadContext.Load` if you do either of the following:

```c#
return null;
return AssemblyLoadContext.Default.LoadByAssemblyName(name);
```

**Review by RikkiGibson — DISMISSED** [review]

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs
Comment (RikkiGibson) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs:135]: Could you please elaborate on what incorrect behavior could arise from the absence of this change? I think we need a test which fails in the absence of this suggested change. Do you have any thoughts on what such a test looks like?

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:455]: Silly question, but did you verify that the test fails when the implementation change in this PR is not present?

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:439]: I'm not certain this test covers the scenario I was thinking of. In the meeting notes I wrote:

> Imagine if they use Newtonsoft in their generator. ServiceHub host might have Newtonsoft loaded. What if the "compiler ALC" given to us by ServiceHub ends up giving us Newtonsoft.Json when we didn't want it?
> 
> We need a unit test for this. An assembly depends on some assembly "Util.dll" that the compiler DOES NOT USE but *would* be loaded by a "compiler ALC". The analyzer needs to reference a different, incompatible version of "util.dll" than the "compiler ALC" would provide. Then we verify that the DirectoryLoadContext loads the analyzer's version of the dependency rather than the version that the "compiler ALC" would give us.

I think this test doesn't cover the scenario described above, since System.Collections.Immutable is an explicit compiler dependency. I'll follow up when I manage to form a coherent thought on this subject.. it seems like the test you created has value, especially if it fails in the absence of the implementation change, but that we probably need more tests.

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs
Comment (genlu) [reply src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:455]: Yes, the test would fail w/o change in this PR.

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs
Comment (genlu) [reply src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:439]: The test I added is based on the error I actually saw when running Roslyn OOP on ServiceHub Core host: we are trying to load MS.CA.CSharp as an analyzer reference, the `DirectoryLoadContext` then delegated the load to `ALC.Default` by returning null in `DirectoryLoadContext.Load`, since MS.CA.CSharp is one of the compiler assemblies on our list.

I can definitely add the test you described in this PR as well, I didn't because I thought it's not directly related to the particular change in this PR.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs
Comment (genlu) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs:74]: I sure can change it to an instance field, but it'd be great if someone can show me what kind of problem it might cause :)

File: src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs
Comment (RikkiGibson) [reply src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:439]: at this point I'm thinking it's OK to merge as long as Jared's feedback is addressed. I am thinking that I will follow up later on and add some more of the gnarly tests. We have another one which looks roughly like the following

- compile an analyzer against net13.
- Use valuetask in the analyzer's implementation.
- find out where valuetask got loaded from. if it's from the default context it's a failure.

Comment (genlu) [reply src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs:439]: OK, added.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs
Comment (jaredpar) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs:135]: Let's identify the different ALC that are in play here: 

1. Compiler ALC: this is the ALC which contains the compiler binaries
2. Default ALC: this is the `AssemblyLoadContext.Default` value 
3. Analyzer ALC: this is the set of ALC loaded for analyzers

In the case of the compiler 1 & 2 are the same. But in the case of VS 1 & 2 are separate with 2 being the service hub process and 1 being the ALC it creates for loading services. 

Lets now consider a case where an analyzer is built against `netstandard1.3`. I chose that target framework because it's different than `netstandard2.0` which is what the compiler typically builds against. That means they will tend to see a different set of references at compile time. Hence we can get into situations where the analyzer has references in its list which don't appear in our `CompilerAssemblySimpleNames` collection. A concrete example of this is `System.Threading.Tasks.Extensions.dll`. That is not included in `CompilerAssemblySimpleNames` but it is a reference that is used by analyzers templates. 

Consider the case of an analyzer that references `System.Threading.Tasks.Extensions.dll` and uses `ValueTask<T>` in one of its methods. When that method is used it will begin with a call to `LoadFromAssemblyName` for `System.Threading.Tasks.Extensions`. That is not in `CompilerAssemblySimpleNames` nor will it be in a known dependency location. Hence it will hit this line and return `null`

The return of `null` means that the load goes back to the default ALC (2 above). Hence between the compiler and VS we end up with two different results. In the compiler it is loaded into the default + compiler ALC as its the same item. In VS the assembly gets loaded into the service hub but does not get loaded into the compiler ALC. 

For types that are never communicated across the compiler API that likely isn't much of a problem. It's an implementation detail of the analyzer so not important if the type isn't accessible in the compiler. I'm guessing if we dig through the facades enough though we will find a case where this results in a type in the compiler public API not being loaded into the compiler ALC. The only guard we have against that today is `CompilerAssemblySimpleNames` and that is an imperfect list. It represents the set of assemblies the compiler knew about at **its** compile time. When an analyzer builds it could have a very different view of the world because types move between assemblies from release to release. It's a case where we can only prove we're safe by exhaustively looking through all TFM and ensuring that types didn't move in such a way that we'd get broken here. I'd prefer to just assume that it probably can happen and guard against it. 

Overall I can't think of any case where we should be loading a managed assembly into the default ALC. It should either be loaded into the analyzer ALC or the compiler ALC. Any time where we fall back to the default ALC we're essentially depending on our host to do the correct thing which isn't reasonable.

Comment (jaredpar) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs:135]: After writing all that part of me wants to suggest we try out deleting `CompilerAssemblySimpleNames`, default to loading into the compiler ALC and then and only then fall back to analyzer ALC if the dependency is in the known dependency list. It's a much bigger change but I'm starting to think it's the correct one. 

Should probably move forward with the direction of this PR though, merge so we unblock VS first though. But still think we need to change this line before merging.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs
Comment (genlu) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs:74]: From the [design note](https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/assemblyloadcontext.md), it is stated that statics are tied to context, so this shouldn't be a problem it seems

**Comment by genlu** [comment]
@jaredpar @RikkiGibson I have made the change to always try to load unknown analyzer dependency in compiler ALC first, but couldn't figure out how to write a test for the `ValueTask<T>` scenario yet. Would you be OK to merge this first to unblock IDE?

**Review by jaredpar — APPROVED** [review]

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs
Comment (RikkiGibson) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs:74]: All I would ask is that we do either the static field or the instance field not both. However this can happen in a follow up.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs
Comment (genlu) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs:74]: That makes sense.

**Pull request merged**
